adapters/oidc/adapter-core/src/main/java/org/keycloak/adapters/OAuthRequestAuthenticator.java

```diff
@@ -398,6 +398,9 @@ private String rewrittenRedirectUri(String originalUri) {
                 StringBuilder redirectUriBuilder = new StringBuilder(url.getProtocol());
                 redirectUriBuilder.append("://"+ url.getAuthority());
                 redirectUriBuilder.append(url.getPath().replaceFirst(rule.getKey(), rule.getValue()));
+                if (url.getQuery() != null && !url.getQuery().isEmpty()) {
+	                redirectUriBuilder.append("?").append(url.getQuery());
```

Comment:
If you have query parameters set to `originalUri` the `?` will be appended twice, right? Perhaps you can check if it is present before adding the parameter? I'm also not sure how this will work if you have fragments in the URL?

Reply:
I testet it locally and when i call it with query parameters the url for the redirect is without the? 
Its because of url.getPath gets only the part without the query parameter. And url.getquery gets the query part without the ?. So when i append it i have to add the ? manually, and it shouldnt be there twice.
But i dont know what you exactliy mean with fragments.

Reply:
But, the issue is already there. And will write a test for it.

Reply:
I see. I was just wondering if creating the `redirectUriBuilder` from the `originalUri` could end up adding the `?` twice.

A test would be nice, thanks. If you can make sure a request URL with and without query parameters works as expected, it should be fine.

Reply:
Sorry for my question here, but, i dont know how to write the test for this case. I looked into your current test but i cant see any test for this function at all. For this function a test need a client with a keycloak json with the redirect rewrite configuration and then to call a keycloak secured site and there to check if the redirect Url is rewirtten. How can i archieve that, can you help me?

Reply:
No problem, this part of the test suite is not so intuitive.

You should be able to add your tests to `DemoServletsAdapterTest`. This class is testing applications using our adapters and if you run it you'll end up at the place you are changing.

You can run the test class from your IDE by making sure you have both `auth-server-quarkus` and `app-server-wildfly` maven profiles enabled. The app server (wildfly) where the application is deployed can be debugged on port `5006`.

Please, let me know if it helps.

In any case, I'm not sure if you want to spend much time on this because we are about to drop our adapters. As mentioned before. So they won't be released.

If you can try using some other capability from the web stack you are using, it is better.

Reply:
I added a test and fix a for a wrong groupId.

Reply:
The rewritten function without parameter is alreadry test.

Reply:
if you approve, is there a way to backport it for version 21.1? I use 21.1.2 and cannot upgrade any further at the moment.

Reply:
Thanks, @mselkecQ. Much appreciated.

I'm pretty sure we are not backporting such changes to those versions. We are backporting changes only to the current latest.

Also, adapters are something we are slowing down so the best we can do, I think, is make it happen in Keycloak 24.